opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerFramework.java

```diff
@@ -0,0 +1,400 @@
+                                prefix,
+                                oldFactory != null ? String.format(". And has replaced existing factory '%s'.", oldFactory.getClass().getCanonicalName()) : ""
+
+                        }
+                );
+            }
+
+            if (factoriesDifferent(factory, oldFactory)) {
+                analyzersInfo.customizations.add("p:" + prefix);
+            }
+        });
+
+        reload();
+    }
+
+    /**
+     * Instruct the AnalyzerGuru to use a given analyzer for a given file extension.
+     * <p>
+     * NOTE: This method calls {@link #reload()} so it may take some time to complete.
+     *
+     * @param extension the file-extension to add
+     * @param factory   a factory which creates the analyzer to use for the given
+     *                  extension (if you pass null as the analyzer, you will disable the
+     *                  analyzer used for that extension)
+     */
+    public void addExtension(String extension, AnalyzerFactory factory) {
+        customizations.add(analyzersInfo -> {
+            AnalyzerFactory oldFactory;
+            if (factory == null) {
+                oldFactory = analyzersInfo.extensions.remove(extension);
+                LOGGER.log(Level.FINER,
+                        "Removing a mapping for suffix {0}{1}",
+                        new Object[]{
+                                extension,
+                                oldFactory != null ? String.format(". And has removed existing factory '%s'.", oldFactory.getClass().getCanonicalName()) : ""
+                        }
+                );
+            } else {
+                oldFactory = analyzersInfo.extensions.put(extension, factory);
+                LOGGER.log(Level.FINER,
+                        "Adding a factory {0} for matching suffix {1}{2}",
+                        new Object[]{
+                                factory.getClass().getCanonicalName(),
+                                extension,
+                                oldFactory != null ? String.format(". And has replaced existing factory '%s'.", oldFactory.getClass().getCanonicalName()) : ""
+                        }
+                );
+            }
+
+            if (factoriesDifferent(factory, oldFactory)) {
+                analyzersInfo.customizations.add("s:" + extension);
+            }
+        });
+
+        reload();
+    }
+
+
+    /**
+     * Free resources associated with all registered analyzers.
+     */
+    public void returnAnalyzers() {
+        getAnalyzersInfo().factories.forEach(AnalyzerFactory::returnAnalyzer);
+    }
+
+    /**
+     * Decode about two factory instances if they are different.
+     *
+     * @param a the first instance
+     * @param b the second instance
+     * @return true if we consider them different, false otherwise
+     */
+    private static boolean factoriesDifferent(AnalyzerFactory a, AnalyzerFactory b) {
```

Comment:
should this class be `Comparable` then ?

Reply:
This is an old code, I can do this in another PR so I don't mix refactoring with actual changes.